share-coordinator/src/test/java/org/apache/kafka/coordinator/share/StateBatchUtilTest.java

```diff
@@ -0,0 +1,472 @@
+                ),
+                -1
+            ),
+
+            new BatchTestHolder(
+                "Handle overlapping batches with different priority.",
+                Collections.singletonList(
+                    new PersisterStateBatch(100, 110, (byte) 0, (short) 1)
+
+                ),  //[(100-115, 0, 1), (121-130, 0, 1)]
+                Arrays.asList(
+                    new PersisterStateBatch(101, 105, (byte) 1, (short) 2),
+                    new PersisterStateBatch(101, 115, (byte) 2, (short) 2),
+                    new PersisterStateBatch(101, 120, (byte) 3, (short) 2)
+                ),  //[(111-123, 2, 2)]
+                Arrays.asList(
+                    new PersisterStateBatch(100, 100, (byte) 0, (short) 1),
+                    new PersisterStateBatch(101, 120, (byte) 3, (short) 2)
+                ),
+                -1
+            ),
+
+            new BatchTestHolder(
+                "Handle overlapping batches within each list with pruning.",
+                Arrays.asList(
+                    new PersisterStateBatch(100, 110, (byte) 0, (short) 1),
+                    new PersisterStateBatch(121, 130, (byte) 0, (short) 1),
+                    new PersisterStateBatch(105, 115, (byte) 0, (short) 1) // overlap with 1st batch
+                ),  //[(100-115, 0, 1), (121-130, 0, 1)]
+                Arrays.asList(
+                    new PersisterStateBatch(111, 119, (byte) 2, (short) 2),
+                    new PersisterStateBatch(116, 123, (byte) 2, (short) 2)  // overlap with first batch
+                ),  //[(111-123, 2, 2)]
+                Arrays.asList(
+                    new PersisterStateBatch(120, 123, (byte) 2, (short) 2),
+                    new PersisterStateBatch(124, 130, (byte) 0, (short) 1)
+                ),
+                120
+            ),
+
+            // complex tests
+            new BatchTestHolder(
+                "Multiple higher state batch updates.",
+                Collections.singletonList(
+                    new PersisterStateBatch(111, 120, (byte) 0, (short) 1)
+                ),
+                Arrays.asList(
+                    new PersisterStateBatch(111, 113, (byte) 0, (short) 2),
+                    new PersisterStateBatch(114, 114, (byte) 2, (short) 1),
+                    new PersisterStateBatch(115, 119, (byte) 0, (short) 2)
+                ),  //[(111-123, 2, 2)]
+                Arrays.asList(
+                    new PersisterStateBatch(111, 113, (byte) 0, (short) 2),
+                    new PersisterStateBatch(114, 114, (byte) 2, (short) 1),
+                    new PersisterStateBatch(115, 119, (byte) 0, (short) 2),
+                    new PersisterStateBatch(120, 120, (byte) 0, (short) 1)
+                ),
+                -1
+            )
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("generator")
+    public void testStateBatchCombine(BatchTestHolder test) {
```

Comment:
I would suggest breaking this up into a few different test cases, organized in whatever way makes sense.

Reply:
will break generator - test will exactly the same